creusot-contracts/src/ghost/ptr_own.rs

```diff
@@ -234,17 +228,174 @@ impl<T: ?Sized> PtrOwn<T> {
+    #[check(ghost)]
+    #[requires(0 <= index && index < self.len())]
+    #[ensures(result.ptr() == self.ptr().thin().offset_logic(index))]
+    #[ensures(*result.val() == self.val()@[index])]
+    pub fn index_ptr_own_ref_ghost(&self, index: Int) -> &PtrOwn<T> {
+        let mut r = self.elements();
+        r.split_off_ghost(index).pop_front_ghost().unwrap()
+    }
+
+    /// Convert a `&mut PtrOwn<[T]>` for a non-empty slice into a `&mut PtrOwn<T>` for the element at the given index.
+    #[check(ghost)]
+    #[requires(0 <= index && index < self.len())]
+    #[ensures(result.ptr() == self.ptr().thin().offset_logic(index))]
+    #[ensures(*result.val() == self.val()@[index])]
+    #[ensures(*(^result).val() == (^self).val()@[index])]
+    #[ensures((^self).ptr() == self.ptr())]
+    #[ensures(forall<k: Int> k != index ==> (^self).val()@.get(k) == self.val()@.get(k))]
+    pub fn index_ptr_own_mut_ghost(&mut self, index: Int) -> &mut PtrOwn<T> {
+        let mut r = self.elements_mut();
+        r.split_off_ghost(index).pop_front_ghost().unwrap()
+    }
+
+    /// Convert `&PtrOwn<T>` to `&PtrLive`
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.ptr() == self.ptr().thin())]
+    #[ensures(result.len()@ == self.len())]
+    pub fn live(&self) -> &PtrLive<T> {
+        panic!("called ghost function in normal code")
+    }
 
-    /// The pointer of a `PtrOwn` is always aligned.
+    /// Split a `&PtrLive` token out of `&mut PtrOwn<T>`.
+    ///
+    /// The original `&mut PtrOwn<T>` is returned so that
+    /// it can be used even during the lifetime of the `&PtrLive`.
+    #[trusted]
     #[check(ghost)]
-    #[ensures(self.ptr().is_aligned_logic())]
-    pub fn ptr_is_aligned_lemma(&self) {}
-
-    /// Opaque wrapper around [`std::ptr::is_aligned_logic`].
-    /// We use this to hide alignment logic by default in `invariant` because it confuses SMT solvers sometimes.
-    /// The underlying property is exposed by [`PtrOwn::ptr_is_aligned_lemma`].
-    #[logic(open(self))]
-    pub fn ptr_is_aligned_opaque(self) -> bool {
-        self.ptr().is_aligned_logic()
+    #[ensures(result.0 == self)]
+    #[ensures(result.1.ptr() == self.ptr().thin())]
+    #[ensures(result.1.len()@ == self.len())]
+    pub fn live_mut(&mut self) -> (&mut Self, &PtrLive<T>) {
+        panic!("called ghost function in normal code")
+    }
+}
+
+/// Evidence that a range of memory is alive.
+///
+/// This evidence enables taking pointer offsets (see [`add_own`](crate::std::intrinsics::add_own) and others)
+/// without ownership (via [`PtrOwn`]) of that range of memory.
+///
+/// It only ever exists as `&PtrLive<T>`: this resource can be duplicated freely,
+/// but its lifetime is bounded by some `&PtrOwn<[T]>` (via `PtrOwn::live`)
+/// or `&mut PtrOwn<[T]>` (via `PtrOwn::live_mut`).
+#[allow(dead_code)]
+#[opaque]
+pub struct PtrLive<T>(PhantomData<T>);
```

Comment:
Instead of only allowing this type to exist under a `&'a`, why not making it `Copy` and annotated with a lifetime?


```suggestion
/// Its lifetime is bounded by some `&PtrOwn<[T]>` (via `PtrOwn::live`)
/// or `&mut PtrOwn<[T]>` (via `PtrOwn::live_mut`).
#[allow(dead_code)]
#[opaque]
#[derive(Clone, Copy)]
pub struct PtrLive<'a, T>(PhantomData<&'a T>);
```